src/set_theory/fixed_points.lean

```diff
@@ -0,0 +1,316 @@
+* `deriv_add_eq_mul_omega_add`: a characterization of the derivative of addition.
+-/
+
+noncomputable theory
+
+universes u v
+
+open function
+
+namespace ordinal
+
+/-! ### Fixed points of type-indexed families of ordinals -/
+
+section
+variables {ι : Type u} {f : ι → ordinal.{max u v} → ordinal.{max u v}}
+
+/-- The next common fixed point, at least `a`, for a family of normal functions. -/
+def nfp_family (f : ι → ordinal → ordinal) (a) : ordinal :=
+sup (list.foldr f a)
+
+theorem nfp_family_eq_sup (f : ι → ordinal → ordinal) (a) :
+  nfp_family f a = sup (list.foldr f a) :=
+rfl
+
+theorem foldr_le_nfp_family (f : ι → ordinal → ordinal) (a l) :
+  list.foldr f a l ≤ nfp_family f a :=
+le_sup _ _
+
+theorem self_le_nfp_family (f : ι → ordinal → ordinal) (a) : a ≤ nfp_family f a :=
+le_sup _ []
+
+theorem lt_nfp_family {a b} : a < nfp_family f b ↔ ∃ l, a < list.foldr f b l :=
+lt_sup
+
+theorem nfp_family_le {a b} : nfp_family f a ≤ b ↔ ∀ l, list.foldr f a l ≤ b :=
+sup_le
+
+theorem nfp_family_monotone (hf : ∀ i, monotone (f i)) : monotone (nfp_family f) :=
+λ a b h, sup_le.2 $ λ l, (list.foldr_monotone hf l h).trans (le_sup _ l)
+
+theorem apply_lt_nfp_family (H : ∀ i, is_normal (f i)) {a b} (hb : b < nfp_family f a) (i) :
+  f i b < nfp_family f a :=
+let ⟨l, hl⟩ := lt_nfp_family.1 hb in lt_sup.2 ⟨i :: l, (H i).strict_mono hl⟩
+
+theorem apply_lt_nfp_family_iff [nonempty ι] (H : ∀ i, is_normal (f i)) {a b} :
+  (∀ i, f i b < nfp_family f a) ↔ b < nfp_family f a :=
+⟨λ h, lt_nfp_family.2 $ let ⟨l, hl⟩ := lt_sup.1 (h (classical.arbitrary ι)) in
+  ⟨l, ((H _).self_le b).trans_lt hl⟩, apply_lt_nfp_family H⟩
+
+theorem nfp_family_le_apply [nonempty ι] (H : ∀ i, is_normal (f i)) {a b} :
+  (∃ i, nfp_family f a ≤ f i b) ↔ nfp_family f a ≤ b :=
+by { rw ←not_iff_not, push_neg, exact apply_lt_nfp_family_iff H }
+
+theorem nfp_family_le_fp (H : ∀ i, monotone (f i)) {a b} (ab : a ≤ b) (h : ∀ i, f i b ≤ b) :
+  nfp_family f a ≤ b :=
+sup_le.2 $ λ l, begin
+  by_cases hι : is_empty ι,
+  { rwa @unique.eq_default _ (@list.unique_of_is_empty ι hι) l },
+  { haveI := not_is_empty_iff.1 hι,
+    induction l with i l IH generalizing a, {exact ab},
+    exact (H i (IH ab)).trans (h i) }
+end
+
+theorem nfp_family_fp (H : ∀ i, is_normal (f i)) (i a) : f i (nfp_family f a) = nfp_family f a :=
+begin
+  unfold nfp_family,
+  rw (H i).sup ⟨[]⟩,
+  apply le_antisymm;
+  refine ordinal.sup_le.2 (λ l, _),
+  { exact le_sup _ (i :: l) },
+  { exact ((H i).self_le _).trans (le_sup _ _) }
+end
+
+theorem apply_le_nfp_family [hι : nonempty ι] {f : ι → ordinal → ordinal} (H : ∀ i, is_normal (f i))
+  {a b} : (∀ i, f i b ≤ nfp_family f a) ↔ b ≤ nfp_family f a :=
+begin
+  refine ⟨λ h, _, λ h i, _⟩,
+  { unfreezingI { cases hι with i },
+    exact ((H i).self_le b).trans (h i) },
+  rw ←nfp_family_fp H i,
+  exact (H i).strict_mono.monotone h
+end
+
+theorem nfp_family_eq_self {f : ι → ordinal → ordinal} {a} (h : ∀ i, f i a = a) :
```

Comment:
Out of pure curiosity, is this an iff? Does `nfp_family f a = a` imply `∀ i, f i a = a`, and if so perhaps that lemma would be nice to have? (I'm happy for this to be a TODO in this file so you can return to it later though)

Reply:
When all functions are normal, it does, but that's just a specific case of `nfp_family_fp`. In fact, this lemma isn't terribly interesting, and I believe I only used it to prove `nfp id = id`.